### PR TITLE
stamp support in storage and protocols

### PR DESCRIPTION
**Pull request by zelig**
#898 
AP https://hackmd.io/2NgPR2CeQICKzWBAH7Mk0Q#

### CODE REVIEW

File: pkg/localstore/export.go [hunk 1]
```diff
@@ -146,7 +149,14 @@ func (db *DB) Import(r io.Reader, legacy bool) (count int64, err error) {
 				case <-ctx.Done():
 				}
 			}
-			stamp := rawdata[:postage.StampSize]
+			stamp := postage.NewStamp(nil, nil)
```
Comment (acud) [inline pkg/localstore/export.go:152]: you can `new(postage.Stamp)`

File: pkg/localstore/localstore_test.go [hunk 2]
```diff
@@ -222,15 +223,15 @@ func TestGenerateTestRandomChunk(t *testing.T) {
 		t.Errorf("first chunk address length %v, want %v", addrLen, 32)
 	}
 	dataLen := len(c1.Data())
-	if dataLen != swarm.ChunkSize+swarm.SpanSize {
+	if dataLen != swarm.ChunkWithSpanSize {
```
Comment (acud) [inline pkg/localstore/localstore_test.go:226]: why is this change needed? these checks are all over the codebase, shouldn't we be PRing such changes to reflect everywhere?

File: pkg/localstore/localstore_test.go [hunk 3]
```diff
@@ -251,7 +252,7 @@ func newRetrieveIndexesTest(db *DB, chunk swarm.Chunk, storeTimestamp, accessTim
 		if err != nil {
 			t.Fatal(err)
 		}
-		validateItem(t, item, chunk.Address().Bytes(), chunk.Data(), storeTimestamp, 0, chunk.Stamp())
+		validateItem(t, item, chunk.Address().Bytes(), chunk.Data(), storeTimestamp, 0)
```
Comment (acud) [inline pkg/localstore/localstore_test.go:255]: by removing these checks you are basically removing guarantees on stamp persistence. may I ask why was this removed?

File: pkg/localstore/localstore_test.go [hunk 4]
```diff
@@ -571,10 +564,11 @@ func TestDBDebugIndexes(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	testIndexCounts(t, 1, 1, 0, 0, 0, 1, 0, indexCounts)
+	// for reference: testIndexCounts(t *testing.T, pushIndex, pullIndex, gcIndex, gcExcludeIndex, pinIndex, retrievalDataIndex, retrievalAccessIndex int, indexInfo map[string]int)
```
Comment (acud) [inline pkg/localstore/localstore_test.go:567]: can be removed

File: pkg/localstore/mode_get.go [hunk 5]
```diff
@@ -50,7 +50,10 @@ func (db *DB) Get(ctx context.Context, mode storage.ModeGet, addr swarm.Address)
 		}
 		return nil, err
 	}
-	return swarm.NewChunk(swarm.NewAddress(out.Address), out.Data).WithPinCounter(out.PinCounter).WithStamp(out.Stamp), nil
+	return swarm.NewChunk(swarm.NewAddress(out.Address), out.Data).
+		WithPinCounter(out.PinCounter).
+		// WithTag(out.Tag).
```
Comment (acud) [inline pkg/localstore/mode_get.go:55]: commented code

File: pkg/postage/stamp.go [hunk 6]
```diff
@@ -102,3 +104,25 @@ func toSignDigest(addr swarm.Address, id []byte) ([]byte, error) {
 	}
 	return h.Sum(nil), nil
 }
+
+func AttachValidStamp(batchStore Storer) func(chunk swarm.Chunk, stampBytes []byte) error {
```
Comment (acud) [inline pkg/postage/stamp.go:108]: I don't understand why this is necessary. You're basically saying: "I don't trust the stamps which are persisted in localstore with the chunk". Can you please elaborate why this is needed?

File: pkg/node/node.go [hunk 7]
```diff
@@ -454,7 +454,7 @@ func NewBee(addr string, swarmAddress swarm.Address, publicKey ecdsa.PublicKey,
 
 	traversalService := traversal.NewService(ns)
 
-	pushSyncProtocol := pushsync.New(p2ps, storer, kad, tagService, pssService.TryUnwrap, logger, acc, accounting.NewFixedPricer(swarmAddress, 1000000000), tracer)
+	pushSyncProtocol := pushsync.New(p2ps, storer, kad, tagService, pssService.TryUnwrap, postage.AttachValidStamp(batchStore), logger, acc, accounting.NewFixedPricer(swarmAddress, 1000000000), tracer)
```
Comment (acud) [inline pkg/node/node.go:457]: Not sure why the protocol should care about postage stamp validity. The necessity of these changes is also not documented in the writeup

File: pkg/shed/db.go [hunk 8]
```diff
@@ -101,35 +101,36 @@ func (db *DB) Put(key, value []byte) (err error) {
...
+		return nil, err
 	}
-	db.metrics.GetCounter.Inc()
 	return value, nil
 }
 
 // Has wraps LevelDB Has method to increment metrics counter.
 func (db *DB) Has(key []byte) (yes bool, err error) {
 	yes, err = db.ldb.Has(key, nil)
+	db.metrics.HasCounter.Inc()
```
Comment (acud) [inline pkg/shed/db.go:118]: you might be incrementing twice here, which is not the purpose of this counter originally. also the changes below do the same. not sure why this is needed

File: pkg/shed/index.go [hunk 9]
```diff
@@ -45,7 +46,41 @@ type Item struct {
...
+
+func (c *Counts) Inc(i int) uint32 {
+	return c.add(i, 1, true)
+}
+func (c *Counts) Dec(i int) uint32 {
+	return c.add(i, 1, false)
+}
+
+func (c *Counts) add(i int, d uint32, add bool) uint32 {
+	count := binary.BigEndian.Uint32(c.Counts[4*i : 4*i+4]) // deserialise
```
Comment (acud) [inline pkg/shed/index.go:76]: this can panic and take the whole node down

File: pkg/shed/index.go [hunk 10]
```diff
@@ -45,7 +46,41 @@ type Item struct {
...
+	if add {
+		count += d
+	} else {
+		count -= d
+	}
+	binary.BigEndian.PutUint32(c.Counts[4*i:4*i+4], count)
```
Comment (acud) [inline pkg/shed/index.go:82]: this too

File: pkg/shed/index.go [hunk 11]
```diff
@@ -189,6 +236,7 @@ func (f Index) Fill(items []Item) (err error) {
 			return fmt.Errorf("decode value: %w", err)
 		}
 		items[i] = v.Merge(item)
+		// items[i] = item.Merge(v)
```
Comment (acud) [inline pkg/shed/index.go:239]: commented code

**Review by acud — CHANGES_REQUESTED** [review]

File: pkg/shed/index.go (hunk 9 above)
Comment (zelig) [reply pkg/shed/index.go:76]: how?

File: pkg/shed/index.go (hunk 10 above)
Comment (zelig) [reply pkg/shed/index.go:82]: how?

File: pkg/node/node.go (hunk 7 above)
Comment (zelig) [reply pkg/node/node.go:457]: push  synced chunks need  to have a valid stamp attached if you dont validate you risk being disconnected and blocklisted for invalid stamp

File: pkg/shed/index.go (hunk 9 above)
Comment (acud) [reply pkg/shed/index.go:76]: i don't see where this field is ever initialised. where does c.Counts get allocated? there is no unit test for this, and where are these offsets coming from?

File: pkg/swarm/swarm.go [hunk 12]
```diff
@@ -120,8 +120,11 @@ type Chunk interface {
 	WithPinCounter(p uint64) Chunk
 	TagID() uint32
 	WithTagID(t uint32) Chunk
-	Stamp() []byte
-	WithStamp([]byte) Chunk
+	Stamp() Stamp
+	WithStamp(Stamp) Chunk
+	Radius() uint8
```
Comment (acud) [inline pkg/swarm/swarm.go:125]: The AP does not specify this interface change. Not sure if this whole AP structure is not generating a ton of overhead to just document everything

File: pkg/pullsync/pullsync_test.go [hunk 13]
```diff
@@ -219,5 +216,6 @@ func newPullSync(s p2p.Streamer, o ...mock.Option) (*pullsync.Syncer, *mock.Pull
 	storage := mock.NewPullStorage(o...)
 	logger := logging.New(ioutil.Discard, 0)
 	unwrap := func(swarm.Chunk) {}
-	return pullsync.New(s, storage, unwrap, logger), storage
+	attachValidStamp := func(ch swarm.Chunk, _ []byte) (swarm.Chunk, error) { return ch, nil }
```
Comment (zelig) [inline pkg/pullsync/pullsync_test.go:219]: naming

File: pkg/swarm/swarm.go [hunk 14]
```diff
@@ -179,10 +190,18 @@ func (c *chunk) TagID() uint32 {
 	return c.tagID
 }
 
-func (c *chunk) Stamp() []byte {
+func (c *chunk) Stamp() Stamp {
 	return c.stamp
 }
 
+func (c *chunk) Radius() uint8 {
```
Comment (zelig) [inline pkg/swarm/swarm.go:197]: comments would be ideal.

     Radius is the PO above which the batch is preserved.

File: pkg/swarm/swarm.go [hunk 15]
```diff
@@ -179,10 +190,18 @@ func (c *chunk) TagID() uint32 {
...
+	return c.radius
+}
+
+func (c *chunk) Depth() uint8 {
```
Comment (zelig) [inline pkg/swarm/swarm.go:201]: Depth returns the batch depth of the stamp - allowed batch size = 2^{depth}

File: pkg/swarm/swarm.go [hunk 16]
```diff
@@ -158,8 +163,14 @@ func (c *chunk) WithTagID(t uint32) Chunk {
 	return c
 }
 
-func (c *chunk) WithStamp(s []byte) Chunk {
-	c.stamp = s
+func (c *chunk) WithStamp(stamp Stamp) Chunk {
```
Comment (zelig) [inline pkg/swarm/swarm.go:166]: comment

Comment (acud) [reply pkg/swarm/swarm.go:166]: added the comments to the interface level

File: pkg/localstore/subscription_push.go [hunk 17]
```diff
@@ -74,7 +74,7 @@ func (db *DB) SubscribePush(ctx context.Context) (c <-chan swarm.Chunk, stop fun
 					}
 
 					select {
-					case chunks <- swarm.NewChunk(swarm.NewAddress(dataItem.Address), dataItem.Data).WithTagID(item.Tag).WithStamp(dataItem.Stamp):
+					case chunks <- swarm.NewChunk(swarm.NewAddress(dataItem.Address), dataItem.Data).WithTagID(item.Tag):
```
[the remark this answers is not in the document]
Comment (acud) [reply pkg/localstore/subscription_push.go:77]: we are, then why did you remove it? :)
I added a test coverage for this case, which was missing :+1:

File: pkg/node/node.go [hunk 18]
```diff
@@ -261,6 +261,7 @@ func NewBee(addr string, swarmAddress swarm.Address, publicKey ecdsa.PublicKey,
 	// register localstore unreserve function on the batchstore before batch service starts listening to blockchain events
 
 	batchStore, err := batchstore.New(stateStore, nil)
+	// batchStore, err := batchstore.New(stateStore, storer.UnreserveBatch)
```
Comment (ralph-pichler) [inline pkg/node/node.go:264]: remove from this PR

File: pkg/pushsync/pushsync.go [hunk 19]
```diff
@@ -153,14 +157,18 @@ func (ps *PushSync) handler(ctx context.Context, p p2p.Peer, stream p2p.Stream)
 // a receipt from that peer and returns error or nil based on the receiving and
 // the validity of the receipt.
 func (ps *PushSync) PushChunkToClosest(ctx context.Context, ch swarm.Chunk) (*Receipt, error) {
-	r, err := ps.pushToClosest(ctx, ch)
+	stamp, err := ch.Stamp().MarshalBinary()
```
Comment (ralph-pichler) [inline pkg/pushsync/pushsync.go:160]: is there a specific reason why the marshal binary is done here and not inside `pushToClosest` directly (which also could do it as it has access to `ch`)?

**Review by ralph-pichler — APPROVED** [review]

File: pkg/pushsync/pushsync.go (hunk 19 above)
Comment (acud) [reply pkg/pushsync/pushsync.go:160]: No clue, I think we can move it in. Good point
